projects/gateway2/proxy_syncer/cla.go

```diff
@@ -69,3 +77,132 @@ func prioritize(ep EndpointsForUpstream) *envoy_config_endpoint_v3.ClusterLoadAs
 	// deprecating the functionality. it's not easy to do as with krt we no longer have gloo 'Endpoint' objects
 	return cla
 }
+
+type uccWithEndpoints struct {
+	Client           krtcollections.UniqlyConnectedClient
+	Endpoints        envoycache.Resource
+	EndpointsVersion uint64
+	endpointsName    string
+}
+
+func (c uccWithEndpoints) ResourceName() string {
+	return fmt.Sprintf("%s/%s", c.Client.ResourceName(), c.endpointsName)
+}
+
+func (c uccWithEndpoints) Equals(in uccWithEndpoints) bool {
+	return c.Client.Equals(in.Client) && c.EndpointsVersion == in.EndpointsVersion
+}
+
+type IndexedEndpoints struct {
+	endpoints krt.Collection[uccWithEndpoints]
+	index     krt.Index[string, uccWithEndpoints]
+}
+
+func (ie *IndexedEndpoints) FetchEndpointsForClient(kctx krt.HandlerContext, ucc krtcollections.UniqlyConnectedClient) []uccWithEndpoints {
+	return krt.Fetch(kctx, ie.endpoints, krt.FilterIndex(ie.index, ucc.ResourceName()))
+}
+
+func NewIndexedEndpoints(logger *zap.Logger, uccs krt.Collection[krtcollections.UniqlyConnectedClient],
```

Comment:
the word endpoint is starting to get confusing in some spots

```suggestion
func NewPerClientEnvoyEndpoints(logger *zap.Logger, uccs krt.Collection[krtcollections.UniqlyConnectedClient],
```